util/progress/progressui/display.go

```diff
@@ -69,53 +88,200 @@ func DisplaySolveStatus(ctx context.Context, c console.Console, w io.Writer, ch
-	width, _ := disp.getSize()
+	var warnings []client.VertexWarning
 	for {
 		select {
 		case <-ctx.Done():
 			return nil, ctx.Err()
 		case <-ticker.C:
+			d.disp.refresh()
 		case ss, ok := <-ch:
-			if ok {
-				t.update(ss, width)
-			} else {
-				done = true
+			if !ok {
+				return warnings, nil
 			}
-		}
 
-		if modeConsole {
-			width, height = disp.getSize()
-			if done {
-				disp.print(t.displayInfo(), width, height, true)
-				t.printErrorLogs(c)
-				return t.warnings(), nil
-			} else if displayLimiter.Allow() {
-				ticker.Stop()
-				ticker = time.NewTicker(tickerTimeout)
-				disp.print(t.displayInfo(), width, height, false)
-			}
-		} else {
-			if done || displayLimiter.Allow() {
-				printer.print(t)
-				if done {
-					t.printErrorLogs(w)
-					return t.warnings(), nil
-				}
-				ticker.Stop()
-				ticker = time.NewTicker(tickerTimeout)
+			d.disp.update(ss)
+			for _, w := range ss.Warnings {
+				warnings = append(warnings, *w)
 			}
+			ticker.Reset(tickerTimeout)
+		}
+	}
+}
+
+type DisplayMode string
+
+const (
+	// DefaultMode is the default value for the DisplayMode.
+	// This is effectively the same as AutoMode.
+	DefaultMode DisplayMode = ""
+	// AutoMode will choose TtyMode or PlainMode depending on if the output is
+	// a tty.
+	AutoMode DisplayMode = "auto"
+	// TtyMode enforces the output is a tty and will otherwise cause an error if it isn't.
+	TtyMode DisplayMode = "tty"
+	// PlainMode is the human-readable plain text output. This mode is not meant to be read
+	// by machines.
+	PlainMode DisplayMode = "plain"
+	// RawJSONMode is the raw JSON text output. It will marshal the various solve status events
+	// to JSON to be read by an external program.
+	RawJSONMode DisplayMode = "rawjson"
+)
+
+// NewDisplay constructs a Display that outputs to the given console.File with the given DisplayMode.
+//
+// This method will return an error when the DisplayMode is invalid or if TtyMode is used but the console.File
+// does not refer to a tty. AutoMode will choose TtyMode or PlainMode depending on if the output is a tty or not.
+func NewDisplay(out console.File, mode DisplayMode, opts ...DisplaySolveStatusOpt) (Display, error) {
+	switch mode {
+	case AutoMode, TtyMode, DefaultMode:
+		if c, err := console.ConsoleFromFile(out); err == nil {
+			return NewConsoleDisplay(c, opts...), nil
+		} else if mode == "tty" {
+			return Display{}, errors.Wrap(err, "failed to get console")
+		}
+		fallthrough
+	case PlainMode:
+		return newPlainDisplay(out, opts...), nil
+	case RawJSONMode:
+		return newRawJSONDisplay(out, opts...), nil
+	default:
+		return Display{}, errors.Errorf("invalid progress mode %s", mode)
+	}
+}
+
+type consoleDisplay struct {
+	t              *trace
+	disp           *ttyDisplay
+	width, height  int
+	displayLimiter *rate.Limiter
+}
+
+// NewConsoleDisplay creates a new Display that prints a TTY
+// friendly output.
+func NewConsoleDisplay(c console.Console, opts ...DisplaySolveStatusOpt) Display {
```

Comment:
I meant private functions for all the displays with specific type. Atm. plain and json are private but console is public.